client/state.py

```diff
@@ -1,131 +1,177 @@
+        try:
+            self.atoms.delete(atom)
+        except ValueError:
+            pass
+
+    def addAtom(self, atom: 'Atom'):
+        self.atoms.update(atom)
+
+    # def __len__(self):
+    #     return self.atoms.kb.length
+
+    def __eq__(self, other: 'State'):
+        return self.atoms == other.atoms  # and self.parent == other.parent and self.last_action == other.last_action
+
+    def __str__(self):
+        state_str = "State " + self.name + "\n"
+
+        state_str += str(self.rigid_atoms)
+
+        state_str += str(self.atoms)
+
+        return (state_str)
+
+    def findBox(self, position):
+        for atom in self.atoms:
+            if atom.name == "BoxAt" and atom.variables[1] == position:
+                return atom
+        return False
+
+    def findBoxGoalDistance(self, name, goalAtom):
+        boxAtom = Atom
+        for atom in self.atoms:
+            if atom.name == "BoxAt" and atom.variables[0] == name:
+                boxAtom = atom
+                break
+        distance = self.getDistance(boxAtom.variables[1], goalAtom["position"])
+        return distance
+
+    def getDistance(self, firstLocation, secondLocation):
+        distance = math.sqrt(math.pow(secondLocation[0] - firstLocation[0], 2) +
+                             math.pow(secondLocation[1] - firstLocation[1], 2))
+        return distance
+
+    def findBoxLetter(self, boxName):
+        for atom in self.rigid_atoms:
+            if atom.name == "Letter" and atom.variables[0] == boxName:
+                return atom
+
+    def findAgent(self, agt):
+        for atom in self.atoms:
+            if atom.name == "AgentAt" and atom.variables[0] == agt:
+                return atom.variables[1]
+
+    def getUnmetGoals(self):
+        metGoals = []
+        unmetGoals = []
+        for goal in self.goals:
+            box = self.findBox(goal['position'])
+            if False == box:
+                unmetGoals.append(goal)
+            else:
+                boxLetter = self.findBoxLetter(box.variables[0])
+                letter = boxLetter.variables[1]
+                if letter != goal['letter']:
+                    unmetGoals.append(goal)
+                else:
+                    metGoals.append(goal)
+        return [unmetGoals, metGoals]
+
+    def getNeithbourGoals(self, position):
+        neighbourLocations = []
+        for atom in self.rigid_atoms:
+            if atom.name == "Neighbour" and position == atom.variables[0]:
+                neighbourLocations.append(atom.variables[1])
+
+        neighbourGoals = []
+        for atom in self.rigid_atoms:
+            if atom.name == "GoalAt":
+                for location in neighbourLocations:
+                    if location == atom.variables[1]:
+                        neighbourGoals.append(atom)
+
+        return neighbourGoals
+
+    def getNeithbourFieldsWithoutGoals(self, position):
+        neighbourLocations = []
+        for atom in self.rigid_atoms:

```

Comment:
You are using two times the same 3 line of codes. You could create a method getNeighbourLocations into the class state